### PR TITLE
Pass dispatch to higher-order middleware and re-implement thunk

### PR DESCRIPTION
**Pull request by acdlite**
See https://github.com/gaearon/redux/issues/113#issuecomment-112603386 for context.

### CODE REVIEW

File: src/middleware/thunk.js [hunk 1]
```diff
@@ -1,10 +1,6 @@
-export default function thunkMiddleware(getState) {
-  return (next) => {
-    const recurse = (action) =>
-      typeof action === 'function' ?
-        action(recurse, getState) :
-        next(action);
-
-    return recurse;
-  };
+export default function thunkMiddleware(getState, dispatch) {
```
Comment (gaearon) [inline src/middleware/thunk.js]: Oh, so if the thunk ain't the first one, `dispatch` will re-start the chain right?

File: src/createDispatcher.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ export default function createDispatcher(store, middlewares = []) {
     }
 
     const finalMiddlewares = typeof middlewares === 'function' ?
-      middlewares(getState) :
+      middlewares({ getState, dispatch }) :
```
Comment (leoasis) [inline src/createDispatcher.js]: Isn't the idea to pass the full dispatch middleware chain instead of the default fn?

Comment (acdlite) [inline src/createDispatcher.js]: D'oh, you're right. I'll update.

File: src/middleware/thunk.js (hunk 1 above)
Comment (acdlite) [inline src/middleware/thunk.js]: Yep!

File: src/createDispatcher.js (hunk 2 above)
Comment (acdlite) [inline src/createDispatcher.js]: Actually I'll update it if/when we merge #119 since there will be conflicts.

**Comment by gaearon** [comment]
Implemented in #195. (Sorry, it was a pain to merge so I just copied the idea.)

**Pull request closed without merging**